### PR TITLE
WIP: prepare script is needed for installing dependencies from git using npm

### PR DESCRIPTION
**Pull request by trusktr**
Hello! 

The `prepare` script is needed in a library in order to make the library installable with `npm` from a git URL in a consuming project's `package.json`.

F.e. A consuming project would have the following to install `rosbag` from git:

```json
"dependencies": {
  "rosbag": "foxglove/rosbag.js",
}
```

However, there's an issue with having both `prepack` and `prepare` scripts when using `npm`: `npm` will run both of them, but when `npm` runs `prepack`, `devDependencies` of the library being installed (f.e. `devDependencies` of `rosbag` while it is being installed from git) will not be available yet, so the build will fail.

The solution to make this work in `npm` is to delete the `prepack` script, and just use `prepare`.

But here's the bad part!

Last I checked, Yarn intentionally [does not run `prepare` scripts](https://github.com/yarnpkg/berry/issues/2305). Instead, Yarn only runs `prepack` scripts, and `devDependencies` are available during `prepack` for `yarn` users, so installing from git with `yarn` works.

Basically Yarn deviated quite significantly from `npm` here, which makes the two totally incompatible with each other when install from git.

If `yarn` must be supported, then adding a `prepare` script and not deleting the `prepack` script won't work. In this case we would need to be sure to change the name of the package and publish it so `npm` users can install from the npm registry.

If `npm` users would like to install the latest from `git` though (while not breaking this for `yarn` users), they would need to clone the project, publish it under their own name, and use a [`package.json` alias](https://stackoverflow.com/a/61246743/454780):

```json
"dependencies": {
    "rosbag": "npm:@my-scope/rosbag@^1.6.1"
}
```

### CODE REVIEW

**Comment by defunctzombie** [comment]
@trusktr Sorry for the delay - we didn't see any notifications for this PR so didn't realize it was hanging out here. We'll get this added today - I agree being able to install directly from github sources is an important workflow.

**Comment by defunctzombie** [comment]
@trusktr Sorry for not commenting on this PR sooner - I didn't get any notifications in my usual channels about it. We currently use yarn for all of our work which includes installing directly from sources in these forks so want to make sure we keep that working.

One option here might be to have prepare and prepack invoke a script that detects yarn vs npm and does the right thing depending on which was invoked.

**Comment by amacneil** [comment]
We are gearing up to publish this package on NPM, so this will become a non-issue for anyone simply wanting to consume this library.

Without doing my own testing, but based on the PR description and also reading yarn + npm docs, it sounds like there is no simple solution that would allow both NPM and yarn users to install the package from git sources and have it automatically built (with devDependencies available during build). It might be possible to add the build step to both prepare and prepack, but wrap prepack in a script which detects and exits under npm. I assume someone on the internet has encountered and solved this issue before so if we are going to fix it, it would be nice to reuse prior art, and use the same pattern across all our libraries and our [template repo](https://github.com/foxglove/template-typescript).

However, given that we will be publishing the package to npm shortly, I'm not convinced that all the above work is necessary. If someone is working from a git fork then they are of course free to edit package.json to rename `prepack` to `prepare`. All other users will just use our public npm module. I therefore suggest closing this PR unless there is another obvious solution I have missed.

**Comment by amacneil** [comment]
This package has now been published to npm as `@foxglove/rosbag`.

If there is a better way to use `prepack` & `prepare` to allow install from git source on both npm and yarn, we're open to suggestions.
